### PR TITLE
Shard a gridfs collection and test working with shards.

### PR DESCRIPTION
**Pull request by manthey**
This adds the ability to ask a gridfs assetstore to shard based on chunk uuid.  As part of this, add a test for a mongo sharding server.

The web_client_admin test is not long run by itself, as recent changes have reduced the timing issues.

### CODE REVIEW

File: girder/utility/gridfs_assetstore_adapter.py [hunk 1]
```diff
@@ -46,13 +47,49 @@
...
+
+    :param collection: a connection to a mongo collection.
     """
     collection.create_index([
         ('uuid', pymongo.ASCENDING),
         ('n', pymongo.ASCENDING)
     ], unique=True)
 
 
+def _setupSharding(collection, keyname='uuid'):
```
Comment (zachmullen) [inline girder/utility/gridfs_assetstore_adapter.py]: The default keyname is 'uuid', but in the adapter it's called 'chunkUuid', and you are using the default value when this is called.

Comment (manthey) [inline girder/utility/gridfs_assetstore_adapter.py]: On the file model, we store the reference in `chunkUuid`, and in the chunk collection, we call it `uuid`.  We are sharding the chunk collection, not the file collection.

Comment (zachmullen) [inline girder/utility/gridfs_assetstore_adapter.py]: Withdrawn!

**Comment by manthey** [comment]
@zachmullen I had to make some changes to work with Mongo 3.0 (which is used in our Circle testing).  Locally, I've tested with Mongo 3.4.  Do we need to test with Mongo 2.6?

**Comment by zachmullen** [comment]
If our CI environment isn't testing 2.6, we probably don't need to here.

**Comment by manthey** [comment]
In which case, this is ready for review.

**Comment by brianhelba** [comment]
Are we still formally supporting MongoDB 2.6?

**Comment by manthey** [comment]
We list 2.6 as the minimum version in the prerequisites.  I'll at least test it manually...

**Comment by mgrauer** [comment]
We discussed it [here](https://github.com/girder/girder/issues/1675) and seemed to converge on dropping support for Mongo 2.6.

If people are good with that as an explicit decision, I can rename that issue to "Drop formal support for Mongo 2.6".

**Comment by manthey** [comment]
I've confirmed that the sharding tests run and pass on Mongo 2.6 (nice to have a Girder vagrant where we can just add `mongodb_version: 2.6.12` to the vars).

File: girder/utility/gridfs_assetstore_adapter.py [hunk 2]
```diff
@@ -93,11 +130,11 @@ def __init__(self, assetstore):
...
         super(GridFsAssetstoreAdapter, self).__init__(assetstore)
         recent = False
-        connectionArgs = (self.assetstore.get('mongohost', None),
-                          self.assetstore.get('replicaset', None))
         try:
             # Guard in case the connectionArgs is unhashable
-            key = connectionArgs
+            key = (self.assetstore.get('mongohost', None),
+                   self.assetstore.get('replicaset', None),
+                   self.assetstore.get('shard', None))
```
Comment (zachmullen) [inline girder/utility/gridfs_assetstore_adapter.py]: `None` is the default for dict.get.

File: tests/base.py [hunk 3]
```diff
@@ -195,12 +197,22 @@ def setUp(self, assetstoreType=None, dropModels=True):
...
                 createGridFsAssetstore(
                 name='Test', db=gridfsDbName,
                 mongohost='mongodb://127.0.0.1:27070,127.0.0.1:27071,'
                 '127.0.0.1:27072', replicaset='replicaset')
+        elif assetstoreType == 'gridfsshard':
+            gridfsDbName = 'girder_test_%s_shard_assetstore_auto' % assetstoreName
+            self.replicaSetConfig = mongo_replicaset.makeConfig(
+                port=27073, shard=True, sharddb=None)
+            mongo_replicaset.startMongoReplicaSet(self.replicaSetConfig)
+            self.assetstore = self.model('assetstore'). \
```
Comment (zachmullen) [inline tests/base.py]: Could we remove the bacskslash and move the function name up to this line for readability

**Comment by zachmullen** [comment]
LGTM other than the minor style nits. Full disclosure, I glossed over the code in `mongo_replicaset.py` -- it seems quite complicated, but since it's just testing code I'm not too worried about that.

File: tests/base.py (hunk 3 above)
Comment (manthey) [inline tests/base.py]: I cleaned up other places where we used the backslash.

**Pull request merged**